src/ACL.cpp

```diff
@@ -141,36 +141,69 @@ QFlags< ChanACL::Perm > ChanACL::effectivePermissions(ServerUser *p, Channel *ch
 	bool write    = false;
 	ChanACL *acl;
 
+	// Iterate over all parent channels from root to the channel the user is in (inclusive)
 	while (!chanstack.isEmpty()) {
 		ch = chanstack.pop();
-		if (!ch->bInheritACL)
+		if (!ch->bInheritACL) {
 			granted = def;
+		}
 
 		foreach (acl, ch->qlACL) {
 			bool matchUser  = (acl->iUserId != -1) && (acl->iUserId == p->iId);
 			bool matchGroup = Group::appliesToUser(*chan, *ch, acl->qsGroup, *p);
+
+			bool applyFromSelf   = (ch == chan && acl->bApplyHere);
+			bool applyFromParent = (ch != chan && acl->bApplySubs);
+
+			// Not using applyFromAny will ignore the context flags for the current ACL
+			bool applyFromAny = applyFromSelf || applyFromParent;
+
+			// Denying traverse is special, because denying it for any parent channel
+			// implicitly means all sub channels are entirely inaccessible as well.
+			// However, we still want to make sure that unchecking "applies to
+			// this channel" for the parent channel is honored.
+			bool applyDenyTraverse = applyFromParent || acl->bApplyHere;
```

Comment:
Wouldn't the described scenario imply that the channel in question is the one whose ACL we currently loop over? I.e. `ch == chan` in which case `acl->bApplyHere == applyFromSelf` and thus `applyFromAny == applyDenyTraverse`?

So my question is: what scenario do you have in mind that requires this special handling?

Reply:
No, consider this scenario:

```
root
  A (traverse denied for THIS channel, but not subs)
    B
      C
```

Try to enter ``C``. The context channel ``chan`` == ``C``. Since the channels are iterated from ``root`` to ``C``, we reach the ACL processing of ``A`` before ``C``.

``applyFromSelf`` will be ``false`` because ``ch`` != ``C``.
``applyFromParent`` will be ``false`` because ``acl->bApplySubs`` == ``false``.
So ``applyFromAny`` will be ``false``. This would lead  to being able to join ``C`` , which is incorrect because ``Traverse`` is **special**.
If we evaluate ``Traverse`` for a single channel to be ``deny``, we expect all children to be inaccessible.

But we evaluate ``applyDenyTraverse`` as ``true`` because of ``acl->bApplyHere`` == ``true`` thus denying and aborting early.

I had to think about this for 15 minutes so I guess I will try to reword the comment

Reply:
Ah, now I see where this is going. Maybe it would be worth it to put this exact example into the comment and then explain with that?